VariantShredding.md

```diff
@@ -25,290 +25,320 @@
-
-# Parquet Layout
+The Parquet columns used to store variant metadata and values must be accessed by name, not by position.
 
-The `array` and `object` fields represent Variant array and object types, respectively.
-Arrays must use the three-level list structure described in [LogicalTypes.md](LogicalTypes.md).
+The series of measurements `34, null, "n/a", 100` would be stored as:
 
-An `object` field must be a group.
-Each field name of this inner group corresponds to the Variant value's object field name.
-Each inner field's type is a recursively shredded variant value: that is, the fields of each object field must be one or more of `object`, `array`, `typed_value` or `variant_value`.
+| Value   | `metadata`       | `value`               | `typed_value` |
+|---------|------------------|-----------------------|---------------|
+| 34      | `01 00` v1/empty | null                  | `34`          |
+| null    | `01 00` v1/empty | `00` (null)           | null          |
+| "n/a"   | `01 00` v1/empty | `13 6E 2F 61` (`n/a`) | null          |
+| 100     | `01 00` v1/empty | null                  | `100`         |
 
-Similarly the elements of an `array` must be a group containing one or more of `object`, `array`, `typed_value` or `variant_value`.
+Both `value` and `typed_value` are optional fields used together to encode a single value.
+Values in the two fields must be interpreted according to the following table:
 
-Each leaf in the schema can store an arbitrary Variant value.
-It contains an `variant_value` binary field and a `typed_value` field.
-If non-null, `variant_value` represents the value stored as a Variant binary.
-The `typed_value` field may be any type that has a corresponding Variant type.
-For each value in the data, at most one of the `typed_value` and `variant_value` may be non-null.
-A writer may omit either field, which is equivalent to all rows being null.
+| `value`  | `typed_value` | Meaning                                                     |
+|----------|---------------|-------------------------------------------------------------|
+| null     | null          | The value is missing; only valid for shredded object fields |
+| non-null | null          | The value is present and may be any type, including null    |
+| null     | non-null      | The value is present and is the shredded type               |
+| non-null | non-null      | The value is present and is a partially shredded object     |
 
-Dictionary IDs in a `variant_value` field refer to entries in the top-level `metadata` field.
+An object is _partially shredded_ when the `value` is an object and the `typed_value` is a shredded object.
 
-For an `object`, a null field means that the field does not exist in the reconstructed Variant object.
-All elements of an `array` must be non-null, since array elements cannote be missing.
+If both fields are non-null and either is not an object, the value is invalid. Readers must either fail or return the `typed_value`.
```

Comment:
```suggestion
If both fields are non-null and either is not an object, the value is invalid.  Reader behavior is undefined[^2]

[^2] The correct thing to do in this situation is rewrite the data file in this scenario (the correct way of rewriting the file can only be determined based on the bug that produced the inconsistency).  Readers are encouraged to error this case to help identify writer bugs as soon as possible.  Erroring might not be appropriate or possible given reader optimizations.  If a reader does not error in this case, then favoring the shredded values would provide more consistent results (i.e. it works across both partial-projection and full reconstruction) but this does not make using the shredded values correct, simply more consistent.  Similarly, and "read side fix" is discouraged, but might be used by specific readers to help their consumers until data can be rewritten.
```

@rdblue @julienledem I think this language might capture all of the concerns across (I think this wording applies to all potential incosistencies).

Reply:
@rdblue  and I also talked about this for a long time and I think I favor the current text. I feel like the additional text adds a bit of confusion around this. 

A shredded reading of a field is always correct since a shredded reader will not be able to check an unshreddeed value for inconsistency, a reader using an un-shredded value when the shredded value is present is always incorrect.

Reply:
> @rdblue and I also talked about this for a long time and I think I favor the current text. I feel like the additional text adds a bit of confusion around this.

We also talked about it in the sync and didn't come to the conclusion.  IIUC @rdblue wanted the error handling language to eliminate the possibility of trying to define an alternative behavior down the road by one specific reader. I don't think we should be mandating this in the spec but I do agree we should be clarifying this won't be relitigated.

> A shredded reading of a field is always correct since a shredded reader will not be able to check an unshreddeed value for inconsistency, a reader using an un-shredded value when the shredded value is present is always incorrect.

I think semantics are important here. I tried to cover this by saying it is "consistent", i.e. it would always provided consistent results to the end user, which is a nice property to have and I agree most implementations should use it if they aren't going to error out.

"correct", I think, is a property of actually returning the the variant data without modification.  Without understanding the bugs that introduced inconsistent shredding, I don't think it is possible say for sure the shredded values are correct.  

The reason why I think it is important to say it is not "correct" is because in other instances (e.g. requiring no overlap between shredded/not shredded fields) to write out of spec files that would not face any issues (as an alternative I think it would OK to have overlapping values as long as they are consistent, then the real question becomes what happens if the values are not actually consistent, in this case I think it would really be very hard to understand what the correct results are).

As an analogy from another part of parquet (and a real world example I've encounted).  Assume a schema like `list<struct<required a int, required b int>>` we do not require readers to check that repetition levels and definition levels are consistent for columns a and b (and many don't).  If they are not equal it is a bug, we can't really say which one, or if either is, "correct" without understanding the bug that introduced the inconsistency.    A "consistent" result would be to always use the left most columns repetition and definition levels.

Reply:
To maybe go a little bit deeper here, I think it is worth enumerating a matrix of potential operations and data inconsistencies (see below).  The main purpose of these examples is to show that:

1.  Requiring using `typed_value` or `failure` pessimizes some cases or requires that different operations will return inconsistent result (making the value of trying to force consistency in some cases have less utility).
2. Preferring `typed_value` in the presence of some inconsistencies sseems like an arbitrary choice simply for the sake of not exposing bugs in the underlying data.

I think the way of rectifying this is either to add more cases that are considered "valid" (e.g. shredded object fields might overlap with fields in value as long as they are consistent) and/or leave the behavior undefined.  My preference would simply be to do the latter.

I think there are three main operations that will be performed with Variant:
1.  Project as primitive non-strict - Return a field as a specific primitive value, if a field mismatches the type return null in its place.
4. Project as primitive strict - Return a field specific primitive value if a field mismatches the type the query is failed.
5. Project as variant - project a field as a variant type allowing for mixed types to be returned (this would also be projecting the top level variant to get the original value).

# Primitive inconsistencies

Assume for all these cases a column is shredded as int32 as its typed value.

## Inconsistency: Both typed_value and value are present for the same cell and the values are consistent with one another (e.g. all values in `value` column are int32 and exactly equal to those in `typed_value`

*Operation*: Project field as int32 non-strict.  In this case engines would prefer to always use typed_value and never need to read `value`

*Operation*: Project field as string not-strict. In this case only `value` column would need to be read but would return consistent results as using `typed_value`

*Operation*: Project field as int32 as strict.  If it is required that a field that can be projected always is (IIUC this isn't currently mandated by the spec), then this operation should fail on statististics as there would be a `non-null` `value` column.  Otherwise this effectively requires the same logic projecting a variant described below.

*Operation*: Project field as string strict.  This would fail based on stastistics from `typed_value`.

*Operation*: Project field as variant.  In this case both `typed_value` and `value` would need to be read and merged.  When merging the reader could chose to always take the non-null value from `typed_value` ignoring any present values in `value` (assuming corresponding values are null).  This would however lead to a strange state where after all values are merged, fewer cells from `value` would be read then are present.  An alternative would be check for consistency between repetition/definition levels, and realize they are both defined and still favor `typed_value`, or check for consistency and continue reading as long as `value` and `typed_value` are consistent.

## Inconsistency: Both type_value and value are present for the same cell and are inconsistent with one another (i.e. `value` contains a string).

*Operation*: Project field as int32 non-strict.  In this case engines would prefer to always use typed_value and never need to read `value`.  int32 value is returned over the string value

*Operation*: Project field as string not-strict. In this case only `value` column would need to be read but would return inconsistent string value in the cells that conflict.

*Operation*: Project field as int32 as strict. Same as data inconsistency above.

*Operation*: Project field as string as strict. Same as data inconsistency above (fail based on statistics).

*Operation*: Project as variant.  Same as data inconsistency above, but if the reader is checking for consistency between values to decide on what to do, as the current spec is written it would arbitrarily choose int32 value (valid alternatives seem like failure or returning the `string` value.


# Object inconsistencies

In this case assume there the object is shredded with a single child field "a" as int32.  Objects overlap with the cases outlined above for consistency/inconsistency so they are not covered here, only somewhat novel cases are discussed below (merging is taken one step further because field overlap either must be detected or duplicate fields could be added when reconstructing the full variant).

## Inconsistency: The `value` object contains a non-object value (e.g. string) in the same cell that has a present value for the shredded column "a".

*Operation*: Project as as variant.  In this case as I read the spec, it is valid for readers to return `{a: <some int value>}`.  Other valid options seem like (use the non-object `value` or fail).

## Inconsistency: The `typed_value` object is marked as non-present but the `value column` is present and contains object.

*Operation*: Project field as object only keeping fields "a" non-strict.  In this case readers would ignore the `value` column based on the spec and return null for the the inconsistency fields.


*Operation*: Project field only keep field "c" as variant.  In this case the reader should be able to ignore `typed_value` (it only contains an object "a" as a shredded column so it can be ignored) and return `{"c": <some value>}` but that is not what the specification states.